### DISCUSSION TITLE
Same service, but different actions

### DISCUSSION THREAD

**Discussion by sebastienbarre** [statelyai/xstate#2330] [General]
To quote the doc: "Actions are fire-and-forget effects. For a machine to be useful in a real-world application, effects need to occur to make things happen in the real world, such as rendering to a screen."

Why is it therefore not possible to create two instances of a machine service with different actions?

As a use case, I want to share the same service between two different UI component, but with different side effects for each UI component...

Thanks

**Comment by davidkpiano** [comment]
> Why is it therefore not possible to create two instances of a machine service with different actions?

The service is the instance. You create different instances by making separate `interpret(...)` calls:

```js
const service1 = interpret(machine); // default actions 
const service2 = interpret(machine.withConfig({
  actions: {/* other actions */}
});
```

**Reply by sebastienbarre** [reply]
Thanks David. But they wouldn't be sharing the same state, right? I'm trying to share the same service between UI components (i.e. the same state), but having different actions being triggered. Meaning if one component sends an event, the other component also reacts to that event, but the side-effects are different (the actions).

In React, I have been using `useMachine()` from two different component, and they appear to react independently. Then again the different between `useInterpret()` and `useMachine()` is unclear to me, maybe there is some magic that makes it work with `useInterpret()`? Is it by virtue of having the same `id`?

**Reply by davidkpiano** [reply]
There is no magic:

- `useMachine(someMachine)` will **create a new service instance** of that machine (nothing is shared - `someMachine` is just a blueprint)
- `useInterpret(someMachine)` will also create a new service instance
- `useService(someService)` will **use that shared service instance**

To solve your use-case, you can use that service with `useService` and execute actions yourself:

```js
const [state, send] = useService(someService);

useEffect(() => {
  state.actions.forEach(action => {
    // execute an effect for each action
    // ...
  });
}, [state]);
```

But now that I think about it, it might be a good idea to have this as a feature of the interpreter 🤔

**Reply by sebastienbarre** [reply]
Thanks for your time. 

I think I see what you mean. In this case I have a state `FOO`, which invokes a `src`, which triggers an action `onFoo` set on `onDone:`, before going to a state `IDLE`. `onFoo` is the action I would want to be different in my `actions:` config object, for different UI components listening to the same service. But since it does not seem possible, I could create an additional indirection state `BAR` as a `target:` to `onDone`, and have that `BAR` state immediately go to `IDLE`. Then in `useEffect()` I could check if `state` == `BAR` and perform my custom action from there...

**Reply by Andarist** [reply]
Just a note - don't check `state` within an effect unless you are absolutely sure that it is what you need.

React might batch renders and by checking `state` within an effect you might "miss" some states. With state machines, it's usually better to encode your logic within the machine itself.

**Reply by sebastienbarre** [reply]
@Andarist thanks, I didn't know that. Is there another way to do it? As you can see, I would like to encode my logic in the machine, but I can't, I need the service to be shared by 2 UI components, and I want the same transition to trigger a different side-effect (aka action) for each component. In other words, the React components are happily tracking the same machine going from state to state, but side effects are up to each component (the state machine services, in this case, are shared, I don't need them to be different; it's the side effect `actions:` I need to be per-component).

So the only information I'm left to work with, is the `state` returned by `useService`...

**Reply by davidkpiano** [reply]
To mitigate the potential issue that @Andarist mentioned, you can subscribe directly:

```js
useEffect(() => {
  return service.subscribe(state => {
    state.actions.forEach(action => {/* execute */})
  }).unsubscribe;
}, [service]);
```

Any sort of helper would basically be doing that, and that's how you can do it as well. For example:

```js
// ...
state.actions.forEach(action => {
  switch (action.type) {
    case 'focus':
      inputRef.current.focus();
      break;
    default:
      break;
  }
});
```

Try that out for now.

**Reply by sebastienbarre** [reply]
@davidkpiano @Andarist thanks a lot for your time and input.

**Reply by sebastienbarre** [reply]
@davidkpiano now that I have played with your example, I'm realizing that `state.actions` is not what I thought it was; I thought it was related to the machine `actions:` setting, but not at all. The only actions I see in `state.actions` while playing with my machine are of type `xstate.start` and `xstate.stop`. I'm having trouble finding documentation about it, except [here](https://xstate.js.org/api/classes/state.html#actions). 

My gut feeling is that if an action `type` is `xstate.stop`, and the action's activity's `src` matches the source in my machine config which would have the action I wish could be unique per component, then this implies the source has terminated successfully. In other words this would be one way to know when the "per-component action" would have been triggered, had it been possible in the first place. So that's probably the time for me to execute my code. I tried it and it seems to work, thanks!

in my machine:
```javascript
    LOGGING_TIMERS: {
        invoke: {
          src: 'logTimers',
          onDone: {
            target: 'IDLE',
            actions: ['onLogTimersCompleted'],
          },
          onError: 'IDLE',
        },
      },
```
In this example, `onLogTimersCompleted` is the action I wish could be different *per component* sharing the same state machine service instance, since in my app two components of mine will have different side effects after the successful completion of `logTimers`.
If I remove the `actions:` from that state, I can instead do this in my component:
```javascript
  useEffect(() => {
    return timeChargingService.subscribe((state) => {
      state.actions.forEach((action) => {
        if (action.type === 'xstate.stop' && action.activity.src === 'logTimers') {
          console.log('timers were logged, time for our additional async stuff');
          ...
        }
      });
    }).unsubscribe;
  }, [timeChargingService]);
```
